java/ql/src/experimental/Security/CWE/CWE-730/RegexInjection.ql

```diff
@@ -0,0 +1,101 @@
+/**
+ * @name Regular expression injection
+ * @description User input should not be used in regular expressions without first being sanitized,
+ *              otherwise a malicious user may be able to provide a regex that could require
+ *              exponential time on certain inputs.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/regex-injection
+ * @tags security
+ *       external/cwe/cwe-730
+ *       external/cwe/cwe-400
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow::PathGraph
+
+/**
+ * A data flow sink for untrusted user input used to construct regular expressions.
+ */
+class RegexSink extends DataFlow::ExprNode {
+  RegexSink() {
+    exists(MethodAccess ma, Method m | m = ma.getMethod() |
+      (
+        m.getDeclaringType().hasQualifiedName("java.lang", "String") and
+        (
+          ma.getArgument(0) = this.asExpr() and
+          (
+            m.hasName("matches") or
+            m.hasName("split") or
+            m.hasName("replaceFirst") or
+            m.hasName("replaceAll")
```

Comment:
You could use a [set literal expression here](https://codeql.github.com/docs/ql-language-reference/expressions/#set-literal-expressions) (might need auto-formatting):
```suggestion
            m.hasName(["matches", "split", "replaceFirst", "replaceAll"])
```
(This would also allow you to get rid of one nesting level of parentheses)